### ISSUE TITLE
zstd dictionary compression for RocksDB and Titan

### ISSUE THREAD

**Issue by yiwu-arbug** [tikv/tikv#8635]
## Development Task

The task is to add config option to enable zstd dictionary compression (https://facebook.github.io/zstd/#small-data), and evaluate its performance under YCSB, sysbench and TPC-C workload.

Background: In rocksdb, data is stored in SST files (default to 8MB currently in TiKV), and inside each SST file, data is divided into data blocks (default to 64KB currently in TiKV but we are considering change it to 4KB). data block is the unit for read IO and compression. However, with the small compression granularity, it is hard to achieve high compression ratio, since the compression algorithm would not have good knowledge of the full data to build a good compression dictionary. To solve the problem, rocksdb introduced zstd dictionary to build compression dictionary per file: https://github.com/facebook/rocksdb/wiki/Dictionary-Compression.

Similarly for Titan, which is our RocksDB plugin with KV separation (https://pingcap.com/blog/titan-storage-engine-design-and-implementation). In Titan the unit for compression is per value, and typical users of Titan would have value size >1KB, but there might be users want to use value size as small as 256B (value even smaller will be inlined in RocksDB LSM). We would like to implement support for dictionary compression for Titan to build compression dictionary per blob file.

Task breakdown:
RocksDB:
- [ ] Add rust-rocksdb interface for rocksdb dictionary compression feature
- [ ] Add tikv config for rocksdb dictionary compression
- [ ] performance test with ycsb/sysbench/tpcc with an IO bounded setup (e.g. with SATA SSD) and both 4k/64k block size scenario.

Titan: see https://github.com/tikv/tikv/issues/5743

Score: 6600

**Comment by kennytm** [comment]
Do those 6600 points include Titan #5743 or just RocksDB alone?

**Comment by yiwu-arbug** [comment]
> Do those 6600 points include Titan #5743 or just RocksDB alone?

including #5743.

**Comment by Rustin-Liu** [comment]
@yiwu-arbug  Could you please modify the description of issue to follow this challenge program [template](https://tidb-community-bots.github.io/challenge-bot/issue-template.html)?

**Comment by yiwu-arbug** [comment]
> @yiwu-arbug Could you please modify the description of issue to follow this challenge program [template](https://tidb-community-bots.github.io/challenge-bot/issue-template.html)?

done

**Comment by Rustin-Liu** [comment]
> done

Thanks for your support.

**Comment by AngleNet** [comment]
@yiwu-arbug  Hi, I have picked the issue. I’m barely new to tikv. I have read some materials list in the descriptions. Here is my initial thoughts about the issue. I will take the top down approach.
1.  Add dict compression to tikv’s kv engine trait, define how tikv uses dict compression from titan and rust-RocksDB
2.Implement dict compression for blob file in titan
3.Add support for dict compression in rust-RocksDB
4.Test titan, tikv with the listed tools with the specified cases.
Feel free to correct me or give me more advice on that. I will write a detailed draft doc on it later. I think the issue will involve tikv, rust-rocksdb and titan, is that correct? Thank you very much!

**Comment by chux0519** [comment]
@AngleNet Hi, I'm interested in this issue as well, Can I join your team and look into this issue together, I think I can do the titan part.

**Comment by AngleNet** [comment]
@chux0519 Welcome! I think there still a position for you. However, I 'm barely new to tikv and its sub projects. It's also possible to drop the challenge for me since I still have daily work to do. It may need very hard working on that. Are you sure to join me? By the way, are you familiar with titan?

**Comment by yiwu-arbug** [comment]
> @yiwu-arbug Hi, I have picked the issue for the challenge . I’m barely new to tikv. I have read some materials list in the descriptions. Here is my initial thoughts about the issue. I will take the top down approach.
> 
> 1. Add dict compression to tikv’s kv engine trait, define how tikv uses dict compression from titan and rust-RocksDB
> 2. Implement dict compression for blob file in titan
> 3. Add support for dict compression in rust-RocksDB
> 4. Test titan, tikv with the listed tools with the specified cases.
> 
> Feel free to correct me or give me more advice on that. I will write a detailed draft doc on it later. I think the issue will involve tikv, rust-rocksdb and titan, is that correct? Thank you very much!

@AngleNet Hi, thank you for your interest on the task. Yes the task involve code changes for tikv, rust-rocksdb and titan. Per your plan:
1. we only need to expose corresponding options to TiKV, so there won't be a lot of change to engine trait. You first need to update rust-rocksdb to expose the options.
2. the Titan part will be a lot of coding, compare to the RocksDB part is just some light coding to expose the option. To flatten the learning curve you can start with the RocksDB change.
3. 4. see below

The order to finish the task should be 
1. add rust-rocksdb interface for the rocksdb dict-compression option
2. add TiKV config for the rocksdb dict-compression option
3. run benchmark (will give you more details once you finish 1 & 2) to test the rocksdb change
4. Implement Titan dict-compression (well, again, there's a lot of coding. Having a teammate to start early on it in parallel to the rocksdb work is also a good idea)
5. run benchmark to test the Titan change (again, more details when you get there).

**Comment by chux0519** [comment]
@AngleNet Yeah I'm pretty sure to join. I'm new to titan, and still browsing the codebase, I think I can finish this part. 
What IM do you prefer, is a telegram group OK? If it is OK for you I will send you a invite link.

**Comment by AngleNet** [comment]
> > @yiwu-arbug Hi, I have picked the issue for the challenge . I’m barely new to tikv. I have read some materials list in the descriptions. Here is my initial thoughts about the issue. I will take the top down approach.
> 
> > 
> 
> > 1. Add dict compression to tikv’s kv engine trait, define how tikv uses dict compression from titan and rust-RocksDB
> 
> > 2. Implement dict compression for blob file in titan
> 
> > 3. Add support for dict compression in rust-RocksDB
> 
> > 4. Test titan, tikv with the listed tools with the specified cases.
> 
> > 
> 
> > Feel free to correct me or give me more advice on that. I will write a detailed draft doc on it later. I think the issue will involve tikv, rust-rocksdb and titan, is that correct? Thank you very much!
> 
> 
> 
> @AngleNet Hi, thank you for your interest on the task. Yes the task involve code changes for tikv, rust-rocksdb and titan. Per your plan:
> 
> 1. we only need to expose corresponding options to TiKV, so there won't be a lot of change to engine trait. You first need to update rust-rocksdb to expose the options.
> 
> 2. the Titan part will be a lot of coding, compare to the RocksDB part is just some light coding to expose the option. To flatten the learning curve you can start with the RocksDB change.
> 
> 
> 
> 3 and 4. see below
> 
> 
> 
> The order to finish the task should be 
> 
> 1. add rust-rocksdb interface for the rocksdb dict-compression option
> 
> 2. add TiKV config for the rocksdb dict-compression option
> 
> 3. run benchmark (will give you more details once you finish 1 & 2) to test the rocksdb change
> 
> 4. Implement Titan dict-compression (well, again, there's a lot of coding. Having a teammate to start early on it in parallel to the rocksdb work is also a good idea)
> 
> 5. add rust-rocksdb interface for the Titan dict-compression option
> 
> 6. add TiKV config for the Titan dict-compression option
> 
> 7. run benchmark to test the Titan change (again, more details when you get there).

Thank you the quick reply. I will post the draft doc as soon as possible.

**Comment by AngleNet** [comment]
> @AngleNet Yeah I'm pretty sure to join. I'm new to titan, and still browsing the codebase, I think I can finish this part. 
> 
> What IM do you prefer, is a telegram group OK? check this out: https://t.me/joinchat/JCjVHAoxXvoqmhes7QTDDQ

Great! I do not have a telegram. WeChat?

**Comment by chux0519** [comment]
@AngleNet my email address: chuxdesign@hotmail.com
you can send your wechat ID via email

**Comment by AngleNet** [comment]
Already sent.

**Comment by AngleNet** [comment]
@yiwu-arbug For the rust-rocksdb part, I think we only need to add `max_dict_bytes` (seems already added)  and `ztsd_max_trains_bytes` the `BlockBasedOptions` in `src/rocksdb_options.rs` along with the FFI (`src/lib.rs`, `librocksdb_sys/crocksdb/c.cc`). Is that right? For the tikv part, we need to add options setter in `bin/tikv-server.rs`  in which the options will be passed to rust-rocksdb to create a fresh rocksdb instance. Seems no change for `engine_trait` ? I still have a problem of how should we export dict-compression configs for titan ? Any sugguestions ? Thank you very much ! 😄 😄

**Comment by yiwu-arbug** [comment]
> @yiwu-arbug For the rust-rocksdb part, I think we only need to add `max_dict_bytes` (seems already added) and `ztsd_max_trains_bytes` the `BlockBasedOptions` in `src/rocksdb_options.rs` along with the FFI (`src/lib.rs`, `librocksdb_sys/crocksdb/c.cc`). Is that right? For the tikv part, we need to add options setter in `bin/tikv-server.rs` in which the options will be passed to rust-rocksdb to create a fresh rocksdb instance. Seems no change for `engine_trait` ? I still have a problem of how should we export dict-compression configs for titan ? Any sugguestions ? Thank you very much ! 😄 😄

yes, that's correct. you only need to add the options. Per titan, I think just adding the same set of options to `TitanCFOptions` would be good. https://github.com/tikv/titan/blob/master/include/titan/options.h#L76

**Comment by yiwu-arbug** [comment]
btw you can join tikv slack (tikv-wg.slack.com) and join the #sig-engine channel for discussion.

**Comment by AngleNet** [comment]
I already joined the slack.

**Comment by yiwu-arbug** [comment]
Task proposal by @AngleNet : https://docs.google.com/document/d/1uMUaF-Hz-DgT2csSQWsdsxIRk3OEgCCqH4fBufKeAlg/edit?usp=sharing

**Comment by vn-ki** [comment]
@AngleNet Are you doing this task? Do you think we can team up for this?

**Comment by yiwu-arbug** [comment]
Changing the scope of this task for RocksDB part only. Titan task is tracked separately with https://github.com/tikv/tikv/issues/5743
